<a href="https://colab.research.google.com/github/allan-gon/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_t_tests_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [0]:
import pandas as pd
import numpy as np
from scipy import  stats

In [0]:
headers = [
           'symboling','normalized-losses','make','fuel-type','aspiration','num-of-doors','body-style','drive-wheels',
           'engine-location','wheel-base','length','width','height','curb-weight','engine-type','num-of-cylinders',
           'engine-size','fuel-system','bore','stroke','compression-ratio','horsepower','peak-rpm','city-mpg','highway-mpg','price'
]
df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data",names=headers)
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [0]:
sample = df.sample(n=10,random_state=30)
mu = df['highway-mpg'].mean()
x_bar = sample['highway-mpg'].mean()
print(f'mu = {mu} , and x bar = {x_bar} they are different but close so i would say that my sample is a reflective of my population')

mu = 30.75121951219512 , and x bar = 32.8 they are different but close so i would say that my sample is a reflective of my population


The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

**Null Hypothesis: The average highway fuel efficiency is 35 mpg**

**Alternative Hypothesis: The average highway fuel efficiency is not 35 mpg**

**Confidence Level: 95%**

In [0]:
stats.ttest_1samp(sample['highway-mpg'], 35,nan_policy='omit')

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

**Fail to reject null hypothesis**                                           
Conclusion the salesman is correct, the average fuel efficiency is about 35 mpg on the highway.

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [0]:
big_samp = df.sample(n=100, random_state=30)
stats.ttest_1samp(big_samp['highway-mpg'], 35,nan_policy='omit')

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

**Reject Null Hypothesis**                                                    
Conclusion the salesman is wrong, the average mpg on the highway is not 35, actually it's less tahn 35

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Because of the sample size. With the sample size of 10 and a seed of 30 we got it was easier to fail to reject the null hypothesis but with more data it became clear that the little inconsistencys were real and not just bad luck when sampling the population

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [0]:
columns = [
           'party','handicapped-infants','water-project',
           'budget','physician-fee-freeze', 'el-salvador-aid',
           'religious-groups','anti-satellite-ban',
           'aid-to-contras','mx-missile','immigration',
           'synfuels', 'education', 'right-to-sue','crime','duty-free',
           'south-africa'
]
cong_df = pd.read_csv("https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data", names = columns)
cong_df.head()

party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [0]:
cong_df.replace("?", np.NaN,inplace=True)

In [0]:
cong_df.replace('y',1,inplace=True)
cong_df.replace('n',0,inplace=True)

In [0]:
dem = cong_df[cong_df['party'] == 'democrat']
rep = cong_df[cong_df['party'] == 'republican']
#rep.head()

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

**Null Hypothesis: Democrats and Republicans don't equally support the bill**

**Alternative Hypothesis: Democrats and Republicans support the bill equally**

**Confidence Interval: 99%**

In [0]:
stats.ttest_ind(rep['handicapped-infants'],dem['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [0]:
print(rep['handicapped-infants'].mean())
print(dem['handicapped-infants'].mean())

0.18787878787878787
0.6046511627906976


**Reject Null Hypothesis:** Republicans and democrats do not support this bill equally, i know this because the p-value is less than 1 - confidence level(.01). In this case, the dems support this bill significantly more than the republicans.

In [0]:
cong_df.columns

Index(['party', 'handicapped-infants', 'water-project', 'budget',
       'physician-fee-freeze', 'el-salvador-aid', 'religious-groups',
       'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 'immigration',
       'synfuels', 'education', 'right-to-sue', 'crime', 'duty-free',
       'south-africa'],
      dtype='object')

**Null Hypothesis: Democrats and Republicans support the bill equally**

**Alternative Hypothesis: Democrats and Republicans don't support the bill equally**

**Confidence Interval: 99%**

In [0]:
#stats.ttest_ind(rep['water-project'],dem['water-project'], nan_policy='omit')#fail to reject null hypothesis
#stats.ttest_ind(rep['budget'], dem['budget'], nan_policy='omit')#here dems support it more, currently looking for reps
stats.ttest_ind(rep['physician-fee-freeze'],dem['physician-fee-freeze'],nan_policy='omit')

Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)

In [0]:
print(rep['physician-fee-freeze'].mean())
print(dem['physician-fee-freeze'].mean())

0.9878787878787879
0.05405405405405406


**Reject Null Hypothesis:** Democrats and Republicans do not support the bill equally. In this case the republicans support the bill significantly more than the republicans.

**Null Hypothesis:** Democrats and Republicans support the bill equally

**Alternative Hypothesis:** Democrats and republicans dont support the bill equally

**Confidence Level:** 90%

In [0]:
stats.ttest_ind(rep['water-project'],dem['water-project'], nan_policy='omit')

Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)

In [0]:
print(rep['water-project'].mean())
print(dem['water-project'].mean())

0.5067567567567568
0.502092050209205


**Fail to reject null hypothesis:** Republicans and democrats do support this bill at almost the same rate. Rep ≈ 50.7 % support, Dem ≈ 50.2% support

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [0]:
def do_2_ttest(df1_col, df2_col, confidence):
  """
  df1_col and df2_col expects a dataframe column like this -> df['names']
  confidence expects a confidence level as a decimal
    for example: if the CL is 99% pass in .99
  """
  alpha = 1 - confidence
  z, p = stats.ttest_ind(df1_col, df2_col, nan_policy='omit')
  if p > alpha:
    return f'Fail to reject null hypothesis\nP = {p} Confidence level = {confidence}'
  else:
    return 'Reject null hypothesis' + '\n' + f'p = {p} Confidence level = {confidence}'
[print(do_2_ttest(rep[column], dem[column], .95)) for column in rep if column != 'party']

Reject null hypothesis
p = 1.613440327937243e-18 Confidence level = 0.95
Fail to reject null hypothesis
P = 0.9291556823993485 Confidence level = 0.95
Reject null hypothesis
p = 2.0703402795404463e-77 Confidence level = 0.95
Reject null hypothesis
p = 1.994262314074344e-177 Confidence level = 0.95
Reject null hypothesis
p = 5.600520111729011e-68 Confidence level = 0.95
Reject null hypothesis
p = 2.3936722520597287e-20 Confidence level = 0.95
Reject null hypothesis
p = 8.521033017443867e-31 Confidence level = 0.95
Reject null hypothesis
p = 2.82471841372357e-54 Confidence level = 0.95
Reject null hypothesis
p = 5.03079265310811e-47 Confidence level = 0.95
Fail to reject null hypothesis
P = 0.08330248490425066 Confidence level = 0.95
Reject null hypothesis
p = 1.5759322301054064e-15 Confidence level = 0.95
Reject null hypothesis
p = 1.8834203990450192e-64 Confidence level = 0.95
Reject null hypothesis
p = 1.2278581709672758e-34 Confidence level = 0.95
Reject null hypothesis
p = 9.9523427

[None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None,
 None]



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [0]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [0]:
### YOUR WORK HERE